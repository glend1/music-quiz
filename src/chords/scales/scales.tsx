import { useState } from "react";
import { TNotes } from "../dictionaryutils/dictionaryutils";
import dStyles from "../dictionaryutils/dictionaryutils.module.css";
import kStyles from "../../canvas/piano/piano.module.css";
import { Piano } from "../../canvas/piano/piano";
import { INote, matchScales, normalizeNote } from "../../notes/notes/notes";
import Image from "next/image";
import * as ReactDOM from "react-dom/client";
import React from "react";
import { ScaleInformation } from "../scaleinformation/scaleinformation";
import { list } from "../../elements/images";

export function Scales({ notes }: TNotes) {
	const [root, setRoot] = useState<string | undefined>();
	if (notes.length > 0) {
		return (
			<>
				<h2>Scales</h2>
				<span className={dStyles.bold}>Select a Scale Root</span>
				<Piano
					clickable={true}
					highlight={[]}
					width={200}
					height={75}
					cb={(e: React.MouseEvent<SVGElement>): void => {
						let element = e.target as SVGElement;
						let alreadySet = element.classList.contains(kStyles.show);
						let highlighted = element.parentNode!.parentNode!.querySelectorAll(
							`.${kStyles.show}`
						);
						if (highlighted.length > 0) {
							highlighted.forEach((el) => {
								el.classList.remove(kStyles.show);
							});
						}
						if (!alreadySet) {
							element.classList.add(kStyles.show);
							setRoot(normalizeNote(element.dataset as INote, 4, true)!.name);
						} else {
							setRoot(undefined);
						}
					}}
				/>
				{root && notes ? (
					<>
						{" "}
						<span className={dStyles.bold}>Scale Names</span>
						{matchScales(notes, root).map((scale, i) => {
							return (
								<div className={dStyles.bubble} key={scale}>
									<span className={dStyles.align}>
										{root} {scale}
									</span>
									<Image
										className="clickable"
										onClick={(e) => {
											let el = (e.target as HTMLElement).parentElement!
												.parentElement!;
											let generated = el.querySelector(".generated");
											if (generated) {
												generated.remove();
											} else {
												let div = document.createElement("div");
												div.className = "generated";
												el.appendChild(div);
												ReactDOM.createRoot(div).render(
													<React.StrictMode>
														<ScaleInformation scale={scale} root={root} />
													</React.StrictMode>
												);
											}
										}}
										src={list}
										alt="Scale Information"
									/>
								</div>
							);
						})}
					</>
				) : null}
			</>
		);
	}
	return null;
}
