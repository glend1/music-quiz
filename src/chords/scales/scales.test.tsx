import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Scales } from "./scales";

describe("scales", () => {
	it("Should render a default component", () => {
		render(<Scales notes={[]} />);
		expect(screen.getByText("Please provide some Notes")).toBeVisible();
	});
	it("Should render scales", async () => {
		const test = render(<Scales notes={["C", "D", "E"]} />);
		await userEvent.click(test.baseElement.querySelectorAll("path")[0]);
		expect(screen.getAllByRole("img").length).toBe(24);
	});
	it("Should return to a default state", async () => {
		const test = render(<Scales notes={["C", "D", "E"]} />);
		await userEvent.click(test.baseElement.querySelectorAll("path")[0]);
		await userEvent.click(test.baseElement.querySelectorAll("path")[0]);
		expect(screen.queryAllByRole("img").length).toBe(0);
	});
	it("Should show/unshow scales", async () => {
		const test = render(<Scales notes={["C", "D", "E"]} />);
		await userEvent.click(test.baseElement.querySelectorAll("path")[5]);
		await userEvent.click(screen.getAllByRole("img")[0]);
		expect(test.baseElement.outerHTML).toMatchSnapshot();
		await userEvent.click(screen.getAllByRole("img")[0]);
		expect(test.baseElement.outerHTML).toMatchSnapshot();
	});
});
