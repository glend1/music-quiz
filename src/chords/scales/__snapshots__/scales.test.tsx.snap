// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scales Should render a default component 1`] = `"<div><h2>Scales</h2><span class="bold">Select a Scale Root</span><div><svg class="clickable" width="200" viewBox="0 0 200 75"><g width="28.571428571428573"><path class="whiteKey key " x="0" data-natural="C" data-octave="lower" rx="2.8571428571428577" ry="3.75" d="M 0 0 v 71.25 q 0 3.75 2.8571428571428577 3.75 h 22.857142857142858 q 2.8571428571428577 0 2.8571428571428577 -3.75 v -71.25 Z"></path><text class="whiteKeyText" x="14.285714285714286" y="65" text-anchor="middle">C</text></g><g width="28.571428571428573"><path class="whiteKey key " x="28.571428571428573" data-natural="D" data-octave="lower" rx="2.8571428571428577" ry="3.75" d="M 28.571428571428573 0 v 71.25 q 0 3.75 2.8571428571428577 3.75 h 22.857142857142858 q 2.8571428571428577 0 2.8571428571428577 -3.75 v -71.25 Z"></path><text class="whiteKeyText" x="42.85714285714286" y="65" text-anchor="middle">D</text></g><g width="14.285714285714286"><path class="blackKey key " x="21.42857142857143" data-sharp="C#" data-flat="Db" data-octave="lower" rx="2.8571428571428577" ry="3.75" d="M 21.42857142857143 0 v 43.125 q 0 3.75 2.8571428571428577 3.75 h 8.571428571428571 q 2.8571428571428577 0 2.8571428571428577 -3.75 v -43.125 Z"></path></g><g width="28.571428571428573"><path class="whiteKey key " x="57.142857142857146" data-natural="E" data-octave="lower" rx="2.8571428571428577" ry="3.75" d="M 57.142857142857146 0 v 71.25 q 0 3.75 2.8571428571428577 3.75 h 22.857142857142858 q 2.8571428571428577 0 2.8571428571428577 -3.75 v -71.25 Z"></path><text class="whiteKeyText" x="71.42857142857143" y="65" text-anchor="middle">E</text></g><g width="14.285714285714286"><path class="blackKey key " x="50" data-sharp="D#" data-flat="Eb" data-octave="lower" rx="2.8571428571428577" ry="3.75" d="M 50 0 v 43.125 q 0 3.75 2.8571428571428577 3.75 h 8.571428571428571 q 2.8571428571428577 0 2.8571428571428577 -3.75 v -43.125 Z"></path></g><g width="28.571428571428573"><path class="whiteKey key " x="85.71428571428572" data-natural="F" data-octave="lower" rx="2.8571428571428577" ry="3.75" d="M 85.71428571428572 0 v 71.25 q 0 3.75 2.8571428571428577 3.75 h 22.857142857142858 q 2.8571428571428577 0 2.8571428571428577 -3.75 v -71.25 Z"></path><text class="whiteKeyText" x="100.00000000000001" y="65" text-anchor="middle">F</text></g><g width="28.571428571428573"><path class="whiteKey key " x="114.28571428571429" data-natural="G" data-octave="lower" rx="2.8571428571428577" ry="3.75" d="M 114.28571428571429 0 v 71.25 q 0 3.75 2.8571428571428577 3.75 h 22.857142857142858 q 2.8571428571428577 0 2.8571428571428577 -3.75 v -71.25 Z"></path><text class="whiteKeyText" x="128.57142857142858" y="65" text-anchor="middle">G</text></g><g width="14.285714285714286"><path class="blackKey key " x="107.14285714285714" data-sharp="F#" data-flat="Gb" data-octave="lower" rx="2.8571428571428577" ry="3.75" d="M 107.14285714285714 0 v 43.125 q 0 3.75 2.8571428571428577 3.75 h 8.571428571428571 q 2.8571428571428577 0 2.8571428571428577 -3.75 v -43.125 Z"></path></g><g width="28.571428571428573"><path class="whiteKey key " x="142.85714285714286" data-natural="A" data-octave="lower" rx="2.8571428571428577" ry="3.75" d="M 142.85714285714286 0 v 71.25 q 0 3.75 2.8571428571428577 3.75 h 22.857142857142858 q 2.8571428571428577 0 2.8571428571428577 -3.75 v -71.25 Z"></path><text class="whiteKeyText" x="157.14285714285714" y="65" text-anchor="middle">A</text></g><g width="14.285714285714286"><path class="blackKey key " x="135.71428571428572" data-sharp="G#" data-flat="Ab" data-octave="lower" rx="2.8571428571428577" ry="3.75" d="M 135.71428571428572 0 v 43.125 q 0 3.75 2.8571428571428577 3.75 h 8.571428571428571 q 2.8571428571428577 0 2.8571428571428577 -3.75 v -43.125 Z"></path></g><g width="28.571428571428573"><path class="whiteKey key " x="171.42857142857144" data-natural="B" data-octave="lower" rx="2.8571428571428577" ry="3.75" d="M 171.42857142857144 0 v 71.25 q 0 3.75 2.8571428571428577 3.75 h 22.857142857142858 q 2.8571428571428577 0 2.8571428571428577 -3.75 v -71.25 Z"></path><text class="whiteKeyText" x="185.71428571428572" y="65" text-anchor="middle">B</text></g><g width="14.285714285714286"><path class="blackKey key " x="164.2857142857143" data-sharp="A#" data-flat="Bb" data-octave="lower" rx="2.8571428571428577" ry="3.75" d="M 164.2857142857143 0 v 43.125 q 0 3.75 2.8571428571428577 3.75 h 8.571428571428571 q 2.8571428571428577 0 2.8571428571428577 -3.75 v -43.125 Z"></path></g></svg></div></div>"`;
